R/utilities.R

```diff
@@ -2236,4 +2236,175 @@ SparseSpearmanCor <- function(X, Y = NULL, cov = FALSE) {
+# @param normMethod Normalization method for the biwig files
+# It can be the name of any quantitative column in the @meta.data
+# or ncells as the number of cells
+# @param cutoff The maximum number of fragment in a given tile
+# @param outdir The output directory for bigwig file
+# Also used as output directory for SlitFragment function
+#' @importFrom rtracklayer import export.bw
+#' @importFrom GenomicRanges seqnames seqlengths GRanges coverage
+#' @importFrom BiocGenerics which
+#' @importFrom S4Vectors match
+#' @importFrom Matrix sparseMatrix rowSums
+CreateBWGroup <- function(groupNamei, availableChr, chromLengths, tiles, normBy, tileSize, normMethod, cutoff, outdir){
+
+  #Read the fragments file associated to the group
+  fragi <- rtracklayer::import(paste0(outdir, "/", groupNamei, ".bed"),format = "bed")
+
+  cellGroupi <- unique(fragi$name)
+
+  #Open the writting bigwig file
+  covFile <- file.path(outdir, paste0(groupNamei, "-TileSize-",tileSize,"-normMethod-",normMethod,".bw"))
+
+  covList <- lapply(seq_along(availableChr), function(k){
+
+    fragik <- fragi[seqnames(fragi) == availableChr[k],]
+    tilesk <- tiles[BiocGenerics::which(seqnames(tiles) %bcin% availableChr[k])]
+
+    if(length(fragik) == 0){
+      tilesk$reads <- 0
+    #If fragments
+    }else{
+
+      #N Tiles
+      nTiles <- chromLengths[availableChr[k]] / tileSize
+      #Add one tile if there is extra bases
+      if (nTiles%%1 != 0) {
+        nTiles <- trunc(nTiles) + 1
+      }
+
+      #Create Sparse Matrix
+      matchID <- S4Vectors::match(mcols(fragik)$name, cellGroupi)
+
+      #For each tiles of this chromosome, create start tile and end tile row, set the associated counts matching with the fragments
+      mat <- Matrix::sparseMatrix(
+        i = c(trunc(start(fragik) / tileSize), trunc(end(fragik) / tileSize)) + 1,
+        j = as.vector(c(matchID, matchID)),
+        x = rep(1,  2*length(fragik)),
+        dims = c(nTiles, length(cellGroupi)))
+
+      #Max count for a cells in a tile is set to cutoff (4)
+      if(!is.null(cutoff)){
+        mat@x[mat@x > cutoff] <- cutoff
+      }
+
+      #Sums the cells
+      mat <- Matrix::rowSums(mat)
+
+      tilesk$reads <- mat
+
+      #Normalization of counts by the sum of readsintss for each cells in group
+      if(normMethod == "ncells"){
+        tilesk$reads <- tilesk$reads / length(cellGroupi)
+      }else if(tolower(normMethod) %in% c("none")){
+      }else{
+        tilesk$reads <- tilesk$reads * 10^4 / sum(normBy[cellGroupi, 1])
+      }
+    }
+
+    tilesk <- coverage(tilesk, weight = tilesk$reads)[[availableChr[k]]]
+
+    tilesk
+
+  })
+
+  names(covList) <- availableChr
+
+  covList <- as(covList, "RleList")
+
+  rtracklayer::export.bw(object = covList, con = covFile)
+
+  covFile
+}
+
+# Helper function for CreateBWGroup
+#' @importFrom S4Vectors match
+'%bcin%' <- function(x, table) S4Vectors::match(x, table, nomatch = 0) > 0
```

Comment:
since this is only used in one place, I don't think we need a helper function. Can use this line itself